src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,144 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
+@SuppressWarnings("unchecked")
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private Object[] elementData;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+    }
+
+    public void rangeCheck() {
+        if (size == elementData.length) {
+            grow();
+        }
+    }
+
+    public void indexCheck(int index) {
+        if (index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
```

Comment:
The `ArrayListIndexOutOfBoundsException` is not defined in the provided files. You need to create this custom exception class as per the task description.